### ISSUE TITLE
grpc_rb_call_run_batch invokes grpc_completion_queue_pluck a seemingly random/arbitrary number of times

### ISSUE THREAD

**Issue by samphippen** [grpc/grpc#10035]
At DigitalOcean we are using gRPC fairly heavily for internal services. While the majority of those services are written in go, we have a couple of core rails monoliths that as a course of necessity we've integrated with our gRPC servers. We've been seeing significantly worse performance with our Ruby clients than our go ones, and so I've been doing some investigation. For us, this increased latency is problematic because we often perform multiple gRPC calls to different services in band in a web request, making web latency spike.

All results here are reported against Ruby 2.4 and revision 73c33541b5eaf6a44e0e14e1382ab19f25538873 of the gRPC library.

To start with, I built a simple rack server that integrates one of our generated Ruby clients and had it sending requests. We use distributed tracing to measure performance and so I was able to break down the request execution as follows:

```
complete request execution time: 100ms
time to first call into gRPC: 0.77ms
gRPC total call execution (this includes remote service execution time): 96ms
remote service execution time: 3ms
```

This gives us a baseline of the total execution time of the gRPC call in Ruby executing in about 30x the time that the server takes to process the request and respond. Network latency is a negligible factor (~1ms).

This lead me to dig in, initially using the excellent [ruby-prof](https://github.com/ruby-prof/ruby-prof) gem to see what was going on. The profile looked like this:

![image](https://cloud.githubusercontent.com/assets/96662/23709219/bbb687e2-03e6-11e7-91a1-47f3fe80758a.png)

From here, I knew I was going to have to dig into the c library, so I cloned down the most recent version of the library and inserted a couple of calls to add tracepoints around various gRPC calls, the git diff being:

```diff
diff --git a/src/ruby/ext/grpc/rb_call.c b/src/ruby/ext/grpc/rb_call.c
index 82d340b..9f3e05a 100644
--- a/src/ruby/ext/grpc/rb_call.c
+++ b/src/ruby/ext/grpc/rb_call.c
@@ -784,6 +784,8 @@ static VALUE grpc_run_batch_stack_build_result(run_batch_stack *st) {
    Only one operation of each type can be active at once in any given
    batch */
 static VALUE grpc_rb_call_run_batch(VALUE self, VALUE ops_hash) {
+  int state;
+  rb_eval_string_protect("Dore::Tracer.start_trace('grpc_batch_setup')", &state);
   run_batch_stack st;
   grpc_rb_call *call = NULL;
   grpc_event ev;
@@ -809,8 +811,11 @@ static VALUE grpc_rb_call_run_batch(VALUE self, VALUE ops_hash) {
   grpc_run_batch_stack_init(&st, write_flag);
   grpc_run_batch_stack_fill_ops(&st, ops_hash);

+  rb_eval_string_protect("Dore::Tracer.finish_trace(false)", &state);
+
   /* call grpc_call_start_batch, then wait for it to complete using
    * pluck_event */
+  rb_eval_string_protect("Dore::Tracer.start_trace('grpc_call_start_batch')", &state);
   err = grpc_call_start_batch(call->wrapped, st.ops, st.op_num, tag, NULL);
   if (err != GRPC_CALL_OK) {
     grpc_run_batch_stack_cleanup(&st);
@@ -819,15 +824,21 @@ static VALUE grpc_rb_call_run_batch(VALUE self, VALUE ops_hash) {
              grpc_call_error_detail_of(err), err);
     return Qnil;
   }
+  rb_eval_string_protect("Dore::Tracer.finish_trace(false)", &state);
+
+  rb_eval_string_protect("Dore::Tracer.start_trace('grpc_completion_queue_pluck')", &state);
   ev = rb_completion_queue_pluck(call->queue, tag,
                                  gpr_inf_future(GPR_CLOCK_REALTIME), NULL);
   if (!ev.success) {
     rb_raise(grpc_rb_eCallError, "call#run_batch failed somehow");
   }
+  rb_eval_string_protect("Dore::Tracer.finish_trace(false)", &state);
+  rb_eval_string_protect("Dore::Tracer.start_trace('grpc_build_result')", &state);
   /* Build and return the BatchResult struct result,
      if there is an error, it's reflected in the status */
   result = grpc_run_batch_stack_build_result(&st);
   grpc_run_batch_stack_cleanup(&st);
+  rb_eval_string_protect("Dore::Tracer.finish_trace(false)", &state);
   return result;
 }
```

This gave the following output in our tracer:

![image](https://cloud.githubusercontent.com/assets/96662/23709326/097dd07a-03e7-11e7-9e9c-bab89b4408e6.png)

From here I was able to determine that the call `rb_completion_queue_pluck` was taking the majority of execution time, so I dug a little further and found that it eventually ends up calling `grpc_completion_queue_pluck`. That call is done inside a callback from `rb_thread_call_without_gvl`, which means I was unable to add tracepoints using our Ruby tracer, because calling back into the VM without the GVL is at best dangerous and at worst going to produce bad data.

So I added a few simple print statements to the loop for each invocation, printing the number of ms that the call to `grpc_completion_queue_pluck` took. I then hit my rack server in a loop, doing two requests to the remote gRPC server per request sent to the rack server, the logs for which are [here](https://gist.github.com/samphippen/14741fc7e19b5bc4ee13932000f7b256)

From this we can see that the server has a "settling in" period where this completion queue polling takes a while and is invoked many times per gRPC call, however after that time we occasionally see significantly slower requests that also invokes the queue many times per gRPC request.

To summarise: it appears that the sort of baseline best case performance for a Ruby gRPC client seems to be around 15ms per request, which would be an acceptable bound if it was highly reproducible, however it seems like this routinely doubles, triples or worse and as such makes it very hard for us to give good SLAs on Ruby services using gRPC. Is there a way that the library can be better invoked by us, or a patch that can be made by the maintainers that will alleviate this issue?

**Comment by apolcyn** [comment]
> At DigitalOcean we are using gRPC fairly heavily for internal services. While the majority of those services are written in go, we have a couple of core rails monoliths that as a course of necessity we've integrated with our gRPC servers. We've been seeing significantly worse performance with our Ruby clients than our go ones, and so I've been doing some investigation.

The large performance disparity between ruby and go is definitely going to be expected. More performance effort is going into java/c#/c++/go than ruby, but these languages also appear to much higher performance potential.

FYI we have some small message benchmarks that show some comparisons between grpc language libraries in
https://performance-dot-grpc-testing.appspot.com/explore?dashboard=5652536396611584. For ruby comparisons, "Streaming secure throughput QPS (8 core client to 8 core server)" and "Unary secure ping pong median latency - Language comparison (in microsec)" are probably relevant.

These benchmarks aren't entirely apples to apples: ruby has only 1000 total outstanding requests while c#/go/java/c++ use 6400 outstanding requests on the server - but note that the scaling between them shows c#/go/java/c++ much greater than ruby. (e.g., with about 6.4 times the number of outstanding calls, c# gets ~40x QPS as ruby).

> rb_completion_queue_pluck was taking the majority of execution time

This is for the most part expected, the basic model is that whenever ruby is making a grpc call, it loses the GIL and calls into the grpc C-library, during this time it repeatedly calls into the C-core library to complete its request (currently with a 20ms timeout on each call).

> Ruby gRPC client seems to be around 15ms per request

This seems unexpectedly high with network latency < 1ms. What are the types/message sizes of the grpc calls? Is this time not just being taken up on the server? Also are the calls all sharing the same client stub? 

In general a larger redesign is probably needed in order for ruby to be in the same ballpark as the c++/go/java/c# libraries, at least in QPS, but the 15ms per call with 1ms latency seems a bit strange. It might be helpful if there's an example client that reproduces.

**Comment by murgatroid99** [comment]
I would like to point out that if you look at the latency benchmark linked above, you can see that our Ruby tests routinely measure an average call latency of about 500 microseconds, so your average latency of 15+ milliseconds is far removed from our own observations.

I would also like to note that the repeated calls to `grpc_completion_queue_pluck` are in a tight loop, without any intervening GVL acquisitions. The overhead of doing that should be negligible compared to the 20ms timeout per call.

**Comment by samphippen** [comment]
Just for my clarity of understanding: does each `grpc_completion_queue_pluck` call result in a network operation?

**Comment by murgatroid99** [comment]
No. Each start_batch call initiates a fixed set of operations; for the most part, a `SEND_*` operation initiates a network write, and a `RECV_*` operation initiates a network read. `grpc_completion_queue_pluck` may cause new reads and writes (e.g. if a TCP read finishes, but more more data is expected, new TCP reads will be started), but for the most part, `grpc_completion_queue_pluck` will simply wait for network events.

**Comment by murgatroid99** [comment]
More to the point, we call `grpc_completion_queue_pluck` with a timeout repeatedly until it returns a non-TIMEOUT event. This should be functionally identical to calling `grpc_completion_queue_pluck` once without a timeout.

**Comment by blowmage** [comment]
Can we get an update on this?

**Comment by murgatroid99** [comment]
One thing I just noticed: the trace above shows that a lot of time is spent in `completion_queue_next` before the server starts processing the request, but it returns almost immediately after the server finishes processing the request. I would normally expect something like that to happen with the first request after the client is created, because that's when gRPC does the TCP handshake, the TLS handshake, the HTTP2 connection establishment, and possibly other connection startup work. But all subsequent requests from the same client should not have that overhead.

Does that pattern occur with all calls? Do you reuse client objects for subsequent requests? If tail latency is a problem, it may be a good idea to just force early connection establishment by making a dummy call with each client you create.

**Comment by samphippen** [comment]
Hi!

Yeah, we worked out that newing up a connection on every request was the cause:

![image](https://cloud.githubusercontent.com/assets/96662/24146123/ab429c1a-0e0a-11e7-9faf-8a3e98953cdf.png)

I think we can close this issue, as this looks like we're using the library in a pretty non-performant way. Thanks for helping everyone ^_^

**Comment by blowmage** [comment]
Glad you got it working Sam!
